pruner/finder.go

```diff
@@ -0,0 +1,109 @@
+package pruner
+
+import (
+	"context"
+	"time"
+
+	"github.com/celestiaorg/celestia-node/header"
+)
+
+// maxHeadersPerLoop is the maximum number of headers to fetch
+// for a prune loop (prevents fetching too many headers at a
+// time for nodes that have a large number of pruneable headers).
+var maxHeadersPerLoop = uint64(512)
+
+// findPruneableHeaders returns all headers that are eligible for pruning
+// (outside the sampling window).
+func (s *Service) findPruneableHeaders(
+	ctx context.Context,
+	lastPruned *header.ExtendedHeader,
+) ([]*header.ExtendedHeader, error) {
+	pruneCutoff := time.Now().UTC().Add(time.Duration(-s.window))
+
+	if !lastPruned.Time().UTC().Before(pruneCutoff) {
+		// this can happen when the network is young and all blocks
+		// are still within the AvailabilityWindow
+		return nil, nil
+	}
+
+	estimatedCutoffHeight, err := s.calculateEstimatedCutoff(ctx, lastPruned, pruneCutoff)
+	if err != nil {
+		return nil, err
+	}
+
+	log.Debugw("finder: fetching header range", "last pruned", lastPruned.Height(),
+		"target height", estimatedCutoffHeight)
+
+	headers, err := s.getter.GetRangeByHeight(ctx, lastPruned, estimatedCutoffHeight)
+	if err != nil {
+		log.Errorw("failed to get range from header store", "from", lastPruned.Height(),
+			"to", estimatedCutoffHeight, "error", err)
+		return nil, err
+	}
+	// ensures genesis block gets pruned
+	if lastPruned.Height() == 1 {
+		headers = append([]*header.ExtendedHeader{lastPruned}, headers...)
+	}
+
+	// if our estimated range didn't cover enough headers, we need to fetch more
+	// TODO: This is really inefficient in the case that lastPruned is the default value, or if the
+	// node has been offline for a long time. Instead of increasing the boundary by one in the for
+	// loop we could increase by a range every iteration
+	headerCount := len(headers)
+	for {
+		if headerCount > int(maxHeadersPerLoop) {
+			headers = headers[:maxHeadersPerLoop]
+			break
+		}
+		lastHeader := headers[len(headers)-1]
+		if lastHeader.Time().After(pruneCutoff) {
+			break
+		}
+
+		nextHeader, err := s.getter.GetByHeight(ctx, lastHeader.Height()+1)
+		if err != nil {
+			log.Errorw("failed to get header by height", "height", lastHeader.Height()+1, "error", err)
+			return nil, err
+		}
+		headers = append(headers, nextHeader)
+		headerCount++
+	}
+
+	for i, h := range headers {
+		if h.Time().After(pruneCutoff) {
```

Comment:
No because `calculateEstimatedCutoff` only estimates the height of the header it expects (given a stable block time) to be the edge of the pruning window, but we need to actually check the timestamps of the headers in the case that we have some variability in block time.

Reply:
This is the whole reason why we had some concerns about using height estimation for pruning.